Remove non-autoloadable classes from preload

jderusse opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | mp
| Tickets       | Fix #44364
| License       | MIT
| Doc PR        | 


When a class is required in a preloaded file, all the type hinted properties have to be auto-loadable. (sees discussion in linked issue)

This could be an issue for classes like `BodyRenderer` that have a property typehinted with an optional dependency.

https://github.com/symfony/symfony/blob/02b40c032865dee88e8f428c08b671896b6fb02c/src/Symfony/Bridge/Twig/Mime/BodyRenderer.php#L27-L39

Sometime, the relation is tricky. For instance `RedisAdapter` uses `RedisTrait` that have a property `$redis` typehinted `RedisClusterProxy` that have a property typehinted `Predis\ClientInterface` => but `Predis` is an optional dependency

When dumping the Preload classMap, this PR now check if all the properties are auto-loadable.

jderusse committed:
Remove non-autoloadable classes from preload

jderusse commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +398,52 @@ class %s extends {$options['class']}
+        }
+        $cache[$class] = true; // prevent recursion
+
+        if (!class_exists($class) && !interface_exists($class, false) && !trait_exists($class, false)) {
```
Why did we checked for `interface_exists` and `trait_exists`?

Shouldn't we preload only classes? If a trait or an interface is used by a class, then it will be preloaded via the class that use it right?

nicolas-grekas replied:
in theory you're right I think, in practice the worst that can happen is these checks being dead code
not having to think about this is why this code is here :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +398,52 @@ class %s extends {$options['class']}
 
+    private function isPreloadable(string $class): ?bool
+    {
+        static $cache = [];
```
no static state please, a property would do fine isn't it?

jderusse replied:
fixed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +398,52 @@ class %s extends {$options['class']}
+
+        // Since PHP 8.1, typehinted properties does not have to be auto-loadable.
+        // Before PHP 8.0, user can not define typehinted properties
+        if (\PHP_VERSION_ID >= 80100 || \PHP_VERSION_ID <= 80000) {
```
the second comparison should use `<` instead of `<=`

stof replied:
typed properties are a PHP 7.4 feature

jderusse replied:
good catch

nicolas-grekas reviewed (commented):
I really like the idea!

I think the current state of the implementation won't work: shouldn't a class that depends on a non-preloadable class also be flagged as non-preloadable?

I also have one fear: in practice, how many classes will this exclude from preloading (especially considering the previous point?) The more typed properties we use, the less classes will be preloaded.

Or maybe we shouldn't be concerned because PHP 8.1 doesn't have the issue, and preloading is also mostly dead since 8.1. We could even consider deprecating our dedicated logic for it.

jderusse commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +399,51 @@ class %s extends {$options['class']}
+        }
+        $this->preloadableCache[$class] = true; // prevent recursion
+
+        if (!class_exists($class) && !interface_exists($class, false) && !trait_exists($class, false)) {
```
beware, switched from `class_exists($class, false)` to `class_exists($class)`

stof commented:
> preloading is also mostly dead since 8.1

@nicolas-grekas what do you mean with this sentence ?

nicolas-grekas commented:
I mean that preloading might not be worth the trouble it is to maintain and setup anymore...

jderusse commented:
> I think the current state of the implementation won't work: shouldn't a class that depends on a non-preloadable class also be flagged as non-preloadable?

Indeed, I missed 3 points:

1. The Preloader browse the classes, properties and methods parameters to discover more classes.

This is now fixed by doing the ~same thing in PHPDumper.

2. The preloader uses `get_declared_classes` to browse more classes

This one is tricky because, in PhpDumper, I can't easily do the same for each discovered class. That would require a tokenizer, or calling `get_declared_classes` in an isolated process => big overhead.

I fixed it by pre-filling the `$preloaded` parameter of the Preloader to prevent it to browse the classes I known to be not pre-loadable.

3. The generated `.preload` file is full of `require_once` included in the Dumped Container 

I have no choice but removing the container from the `.preload.php` file. (in my case, the code preloaded 2300 files before, and 1900 after that change)

derrabus commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +406,108 @@ class %s extends {$options['class']}
+            if ($t->isBuiltin()) {
+                continue;
+            }
+            $class = $t instanceof \ReflectionNamedType ? $t->getName() : (string) $t;
```
```suggestion
            $class = $t->getName();
```

On PHP 7.2 - 8.1, `ReflectionUnionType`, `ReflectionIntersectionType` and `ReflectionNamedType` are the only subtypes of `ReflectionType`. If we really get something else here, we should not automatically assume, that it's stringable.

jderusse replied:
the abstract `ReflectionType` class implements `Stringable`.
`+` this code is not called for PHP >= 8.1

derrabus replied:
Let me be more precise: We should not assume that casting an instance of an unknown `ReflectionType` subtype that might be added some time in the future to string will yield a class.

Also, by calling `->isBuiltin()` you kind-of already assume that you have a `ReflectionNamedType` instance. Maybe we can move the safeguard there:

```php
if (!$t instanceof \ReflectionNamedType || $t->isBuiltin()) {
    continue;
}
$class = $t->getName();
```

fabpot commented:
This one is quite important as currently, Symfony 6.0 is broken when using pre-loading.

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -325,24 +326,31 @@ class %s extends {$options['class']}
 }
 
 require $autoloadFile;
+if (\PHP_VERSION_ID >= 80100 || \PHP_VERSION_ID < 70400) {
```
`\PHP_VERSION_ID < 70400` is useless here. the whole preloading feature is requiring PHP 7.4+ (as PHP 7.3 does not have preloading at all in PHP)

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +406,111 @@ class %s extends {$options['class']}
+            }
+        }
+
+        // code bellow does not prevent the class to be preloaded, but is here to warm the preloadableCache
```
typo: `below`, not `bellow`: https://prowritingaid.com/grammar/1000167/Bellow-vs-below%E2%80%94what-is-the-difference

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -398,6 +406,111 @@ class %s extends {$options['class']}
+        }
+
+        // code bellow does not prevent the class to be preloaded, but is here to warm the preloadableCache
+        foreach ($r->getMethods(\ReflectionMethod::IS_PUBLIC) as $m) {
```
isn't this also true for protected methods ? They are also involved in inheritance checks.

fabpot commented:
See https://github.com/symfony/symfony/pull/44494 for an alternative

fabpot commented:
Closing in favor of #44494

jderusse commented:
reopening: Removing typehinted property in `symfony/symfony: 6.0` is a workaround for our internal classes, but the bug still exists for other dependencies.

The current implementation of the preloads script discovers all classes used by the end-user. Every third-party package/bundle with type hinted property will hit the same bug.

Worst case, this PR should be merged in 6.1 and type-hint re-added.

fabpot commented:
This one should be rebased on 6.1 then.

nicolas-grekas commented:
> > The preloader uses get_declared_classes to browse more classes

> This one is tricky because, in PhpDumper, I can't easily do the same for each discovered class. That would require a tokenizer, or calling get_declared_classes in an isolated process => big overhead.

Could we fix this by registering an autoloader that only collects autoloaded symbols before the real autoloader?

> > The generated .preload file is full of require_once included in the Dumped Container

> I have no choice but removing the container from the .preload.php file. (in my case, the code preloaded 2300 files before, and 1900 after that change)

You mean in the initializer that is called on the first call to `set()` since #44010, right?
We might be able to fix this by not calling `set()` but listing the files/classes in the preload script instead, isn't it?

nicolas-grekas commented:
What about another alternative: deprecate generating the preload script at all?

Since 8.1, preloading [provides almost no benefits](https://www.php.net/releases/8.1/) to me.

Preloading is (was?) also sensitive to "over-preloading" (aka preloading too many classes, making it under-perform due to hash table overhead), so that any benefits provided by preloading might easily be canceled by having it not properly configured. And since configuring it is quite difficult, the gain might not be reachable in practice.

Also, I guess 8.1 is able to cache the same class/symbol for different paths, while preloading cannot.

I'd be interested in knowing @nikic's thoughts on the topic!

Or we could make the preloading script extremely simpler, eg preloading only composer+some very basic stuffs.

nicolas-grekas commented:
BTW, would it possible to turn that fatal error into a warning on PHP 8.0, as a bugfix?

nikic commented:
> BTW, would it possible to turn that fatal error into a warning on PHP 8.0, as a bugfix?

Nope, the fatal error is required for correctness in 8.0.

As to preloading in general, the usefulness is greatly diminished with 8.1, but I don't really have data on whether there are any cases where it still makes sense.

wouterj commented:
As we've added the type to properties again in the 6.1-dev branch, what is needed to finish this PR? Is there anything I (or the community) can do to help move this forward?

nicolas-grekas commented:
Replaced by #45377

Thank you very much for providing this patch @jderusse, it's been critical to make an informed decision on the topic! :pray:
